### PR TITLE
#9420 - Parsing for precipitation values fails for some weather stat files

### LINKED ISSUE #9420: Parsing for precipitation values fails for some weather stat files
Issue overview
--------------
User file fails with this cryptic error:
```
   ************* Beginning Simulation
   ** Severe  ** invalid stoi argument
   ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
```

There are [two places where `std::stoi` is used](https://github.com/NREL/EnergyPlus/blob/463ec66491ea8a9764668e600abbc67761e45117/src/EnergyPlus/OutputReportTabular.cc#L5928-L5969) to parse lines from the weather stat file for reporting monthly total and max precipitation.

The line being parsed when it fails is this header under ` - Wind Frequency (m/s) by Direction {Interval 11.25 deg from displayed deg) [%]`:
` 	           	<=.5 	<=2  	<=4  	<=6  	<=8  	<=10 	<=12 	<=14 	<=16 	<=18 	>18  	Total	`

This is a newer block that does not exist in the stat files distributed with EnergyPlus.

The root of the problem is [here in `parseStatLine`](https://github.com/NREL/EnergyPlus/blob/463ec66491ea8a9764668e600abbc67761e45117/src/EnergyPlus/OutputReportTabular.cc#L5415-L5420) which assumes that any line containing "Max Hourly" will be for ` - Monthly Statistics for Liquid Precipitation [mm]`. The stat file for this defect has two lines with "Max Hourly" one for liquid precipitation and one for snow depth.

`parseStatLine` also assumes that there will be only 2 occurences of "Total" in the entire stat file, once for `Average Hourly Statistics for Total Sky Cover` and once for liquid precipitation. The stat file for this defect has 4 occurences of "Total".

1. `parseStatLine` needs to be more robust to ensure that values are being pulled from the correct block of statistics. The loop that calls `parseStatLine` needs to be modified to search for " - Monthly Statistics for Liquid Precipitation [mm]" first, then look for "Total" and "Max Hourly".
2. `std::stoi` is fragile and needs to be protected to produce a useful error message, or ignore exceptions and move on to the next statistic, or be replaced with a more robust function such as `UtilityRoutines::ProcessNumber`.

### Workaround
Rename the stat file so that EnergyPlus cannot find it. The simulation will complete, but the Weather Statistics File subtable of the Climatic Data Summary table output will be empty.

### Details
Some additional details for this issue (if relevant):
 - Platform Windows 10
 - Version of EnergyPlus v22.1
 - Private email and helpdesk ticket number nnnn

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file attached: [9420-weather error.zip](https://github.com/NREL/EnergyPlus/files/8664660/9420-weather.error.zip)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #9420 
 - More robustly locate the precipitation data. In particular for stat files from climate.onebuilding.org and similar ones.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/OutputReportTabular.unit.cc [hunk 1]
```diff
@@ -10128,3 +10129,53 @@ TEST_F(SQLiteFixture, OutputReportTabularMonthly_CurlyBraces)
         EXPECT_TRUE(false) << "Missing braces in monthly table for : " << colHeader;
     }
 }
+
+TEST_F(SQLiteFixture, StatFile_TMYx)
+{
+    // Test for #9400 and #9420
+    state->files.inStatFilePath.filePath =
+        configured_source_directory() / "tst/EnergyPlus/unit/Resources/USA_IL_Chicago.OHare.Intl.AP.725300_TMYx.stat";
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/OutputReportTabular.unit.cc:10125]: New unit test.

Before fix:

```
[ RUN      ] SQLiteFixture.StatFile_TMYx
unknown file: Failure
C++ exception with description "stoi" thrown in the test body.
[  FAILED  ] SQLiteFixture.StatFile_TMYx (161 ms)
```

File: src/EnergyPlus/OutputReportTabular.hh [hunk 2]
```diff
@@ -606,7 +606,8 @@ namespace OutputReportTabular {
                        bool &desConditionlinepassed,
                        bool &heatingDesignlinepassed,
                        bool &coolingDesignlinepassed,
-                       bool &isKoppen);
+                       bool &isKoppen,
+                       bool &insideLiquidPrecipitation);
```
Comment (jmarrec) [inline src/EnergyPlus/OutputReportTabular.hh:610]: new flag to make sure you're in the right block

File: src/EnergyPlus/OutputReportTabular.cc [hunk 3]
```diff
@@ -5411,13 +5412,16 @@ void parseStatLine(const std::string &lineIn,
...
+        lineType = StatLineType::MonthlyPrec;
+    } else if (insideLiquidPrecipitation && has(lineIn, "Max Hourly")) {
         lineType = StatLineType::MaxHourlyPrec;
-    } else if (has(lineIn, "Total")) {
-        if (!has(lineIn, "Statistics for Total Sky Cover")) {
-            lineType = StatLineType::MonthlyPrec;
-        }
+        insideLiquidPrecipitation = false;
     }
+
```
Comment (jmarrec) [inline src/EnergyPlus/OutputReportTabular.cc:5424]: Identify right block

**Comment by Myoldmopar** [comment]
CI looks very happy here, looking over the code changes now.

File: scripts/dev/find_byref_bool_override.py [hunk 4]
```diff
@@ -337,7 +337,8 @@
             "isKoppen",
             "heatingDesignlinepassed",
             "coolingDesignlinepassed",
-            "desConditionlinepassed"
+            "desConditionlinepassed",
+            "insideLiquidPrecipitation",
```
Comment (Myoldmopar) [inline scripts/dev/find_byref_bool_override.py:341]: New bool byref approved-list entry.  Fine.

File: src/EnergyPlus/OutputReportTabular.cc [hunk 5]
```diff
@@ -5362,7 +5362,8 @@ void parseStatLine(const std::string &lineIn,
                    bool &desConditionlinepassed,
                    bool &heatingDesignlinepassed,
                    bool &coolingDesignlinepassed,
-                   bool &isKoppen)
+                   bool &isKoppen,
+                   bool &insideLiquidPrecipitation)
```
Comment (Myoldmopar) [inline src/EnergyPlus/OutputReportTabular.cc:5366]: Passing another boolean here makes me squirm a little, but I recognize that this is _not_ something that is going to be changing almost ever, so this is ... OK :smile:

File: src/EnergyPlus/OutputReportTabular.cc (hunk 3 above)
Comment (Myoldmopar) [reply src/EnergyPlus/OutputReportTabular.cc:5424]: Got it.  Catching the statistics line ahead in the IF block and setting the flag.  And splitting the two "Total" lines is fine here.  Looks a bit cleaner.

File: tst/EnergyPlus/unit/OutputReportTabular.unit.cc [hunk 6]
```diff
@@ -8229,41 +8230,29 @@ TEST_F(SQLiteFixture, OutputReportTabular_EndUseBySubcategorySQL)
...
+    auto getLineType = [](const std::string statLine) -> StatLineType {
+        StatLineType lineTypeReturn = StatLineType::Initialized;
+        bool desCondLinePassed = false;
+        bool htgDesignLinePassed = false;
+        bool clgDesignLinePassed = false;
+        bool isKoppen = false;
+        bool insideLiquidPrecipitation = false;
+        parseStatLine(statLine, lineTypeReturn, desCondLinePassed, htgDesignLinePassed, clgDesignLinePassed, isKoppen, insideLiquidPrecipitation);
+        return lineTypeReturn;
+    };
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/OutputReportTabular.unit.cc:8243]: Reusable lambda, a nice touch.

File: tst/EnergyPlus/unit/OutputReportTabular.unit.cc [hunk 7]
```diff
@@ -10128,3 +10117,53 @@ TEST_F(SQLiteFixture, OutputReportTabularMonthly_CurlyBraces)
...
+    state->dataOutRptTab->unitsStyle = OutputReportTabular::UnitsStyle::InchPound;
+    state->dataOutRptTab->unitsStyle_SQLite = OutputReportTabular::UnitsStyle::InchPound;
+
+    SetPredefinedTables(*state);
+
+    FillWeatherPredefinedEntries(*state);
+
+    // Enable the ClimaticDataSummary report, in a future-proof way
+    auto &predefReports = state->dataOutRptPredefined->reportName;
+    auto it = std::find_if(predefReports.begin(), predefReports.end(), [](const auto &s) { return s.name == "ClimaticDataSummary"; });
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/OutputReportTabular.unit.cc:10142]: Nice, finding this report dynamically is great!

**Review by Myoldmopar — APPROVED** [review]
This is awesome.  Thanks @jmarrec !

**Comment by Myoldmopar** [comment]
Confirmed the stoi error on develop, and verified this fixes it with a clean simulation result and error file.  Thanks @jmarrec, merging.

**Pull request merged**